agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -149,6 +169,184 @@ impl DeriveSchema for Stage {
             }))
         }
 
+        fn graph_lookup_derive_schema(
+            graph_lookup: &GraphLookup,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            // it is a bit annoying that we need to clone these Strings just to do a lookup, but I
+            // don't think there is a way around that.
+            let from_field = Namespace(state.current_db.clone(), graph_lookup.from.clone());
+            let mut from_schema = state
+                .catalog
+                .get(&from_field)
+                .ok_or_else(|| Error::UnknownReference(from_field.to_string()))?
+                .clone();
+            if let Some(ref depth_field) = graph_lookup.depth_field {
+                let depth_field_schema = Schema::Atomic(Atomic::Long);
+                insert_required_key_into_document(
+                    &mut from_schema,
+                    depth_field_schema,
+                    vec![depth_field.clone()],
```

Comment:
```suggestion
                    depth_field.as_str().split('.').map(|s| s.to_string()).collect(),
```
This might not be perfect with rustfmt/references but I think this is the right idea. I tested it and it looks like if you have a document field in the from collection and the depth_field is nested, it will treat the field name as a path rather than a literal. It is ugly but here is a demonstration from my local mongo:
```
derivation> db.aggregate([
...   {$documents: [{_id: 1, "foo": 1,},]},
...   {"$graphLookup": {"from": "bar", "startWith": "$foo", "connectFromField": "foo", "connectToField": "baz", "depthField": "DEPTH", "as": "arr"}}
... ])
[
  {
    _id: 1,
    foo: 1,
    arr: [ { _id: 0, baz: 1, doc: { a: 0 }, DEPTH: Long('0') } ]
  }
]
derivation> db.aggregate([
...   {$documents: [{_id: 1, "foo": 1,},]},
...   {"$graphLookup": {"from": "bar", "startWith": "$foo", "connectFromField": "foo", "connectToField": "baz", "depthField": "doc.DEPTH", "as": "arr"}}
... ])
[
  {
    _id: 1,
    foo: 1,
    arr: [ { _id: 0, baz: 1, doc: { a: 0, DEPTH: Long('0') } } ]
  }
]
```

Reply:
I assume this will also hold for the `_as` fields in normal $lookups

Reply:
Yes, good point